1-Draft/RFC00XX-PredIctive-IntelliSense.md

```diff
@@ -0,0 +1,359 @@
+additional providers.
+
+Predictions may be displayed in either **InlineView** or **ListView** depending on the user’s
+preference. Below, predictions are displayed with InlineView.
+
+![image](./media/pred-wlc.png)
+
+Below, predictions are displayed with a dropdown in ListView.
+
+![image](./media/pred-listview.png)
+
+## Motivation
+
+Tab Completion has accelerated the success of both new and experienced PowerShell users. New users
+get the benefit of discovery; seeing available cmdlets and parameters as options while interactively
+typing. Experienced users receive the benefit of acceleration; typing less while using the **<tab>** key
+to quickly complete a command.
+
+The increasing amount of technology translates to an increase in cmdlets and full command
+complexity. Predictive IntelliSense is an addition to the concept of Tab Completion; helping the
+user discover, build and edit full commands based on the user’s history or additional plugins.
+
+## Release plan
+
+- Proposed 7.1 GA
+  - Currently in PowerShell 7.1 PReview 7
+  - Predictive IntelliSense Whole Line Completion
+  - Historical Predictions
+  - Extension Framework
+- Future PSReadLine previews
+  - Predictive IntelliSense Whole Line Completion + ListView
+  - Support for additional providers using extension framework
+  - Previews begin in October with PSReadLine 2.2.0-beta1
+
+## Goals/Non-Goals
+
+ Goals:
+
+- Provide suggestions from PSReadLine history
+- The user will be able to enable and disable Predictive IntelliSense
+- The user will be able to change between InlineView and ListView
+- The user will have keyboard shortcuts to navigate and edit a prediction
+- The user will register additional providers when desired
+- The user may customize the color of a prediction to support accessibility needs
+- The user will be able to search history for similar predictions
+
+Non-goals:
+
+- Due to limit use, PowerShell Session history is not part of the historical suggestion
+- VSCode implementation. We agree this is important for a consistent experience and are working to
+  bring this feature in a future release
+
+## Specification
+
+The proposal is to add Predictive IntelliSense and Extension model to improve the user interactive experience.
+
+### Key Design Considerations Predictive IntelliSense
+
+Predictive IntelliSense displays a best match command completion from command history. The
+prediction may change if a better match is found as the user types additional information. The
+predicted command may be accepted, edited, or ignored.
+
+![Prediction - Historical](./media/pre-wlc.png)
+
+### Install and Remove Predictive IntelliSense
+
+Predictive IntelliSense is a part of the PSReadLine module. Updating or installing the current beta
+release from PSGallery will provide the Predictive IntelliSense feature.
+
+The current release is PSReadLine 2.1.0 Beta 2:
+
+```powershell
+Install-Module PSReadLine -RequiredVersion 2.1.0-beta2 -AllowPrerelease
```

Comment:
This doesn't work on WPS 5.1. What's the WPS equivalent?

Thanks.

Reply:
The install command or the PSReadLine version?

I believe PSReadLine 2.1 supports WinPS 5.1.

The command may not work if your `PowerShellGet` version is too old. You may need to run `Install-Module PowerShellGet` first to get a version that supports `Install-Module -AllowPrerelease`

Reply:
The install command.

It was several undocumented and painful-to-discover steps. The initial errors complain that the version isn't valid. Find-Module agrees and spills a sea of red ink. It was a horrible user experience.

Your normal admin-scripter will stay away. I would have done so if I really had a choice. Unfortunately, I had a long-term crash bug in PSreadline that I wanted the fix to.

Reply:
What's the result of `gmo PowerShellGet -List` on your machine?

Reply:
Now it's

> ModuleType Version    Name
> ---------- -------    ----
> Script     2.2.5      PowerShellGet
> Script     1.0.0.1    PowerShellGet

Before, it was just version 1.0.0.1.

Reply:
Yeah unfortunately because the 1.0.0.1 version is the one that's in Windows it's the one that many users have, but it's not at all easy to push an upgrade of it through Windows like that (improving the UX at the cost of possibly breaking existing workflows in Windows doesn't fly). All the UX improvements come in later versions, so there's the necessary step of using the version in Windows to bootstrap the new version. At least in Windows PowerShell. PowerShell 7 ships with the newer version.

Reply:
I understand.

But you will always find me pushing the "cause" of the admin-scripter. S/he's the person who has to sit down and bang through this trying to figure out "WTH?".  :-)

Reply:
I think there is a separate discussion outside of the RFC to have PowerShellGetv3 in WinPS in the future.